### PR TITLE
Generate HTML output files of test runs

### PR DESCRIPTION
**Pull request by jGleitz**
I spend the day fixing #39 once and for all. (Okay, maybe it was more for fun than to actually fix this, but…)

This pull introduces the `outputFile` JVM argument, which lets you specify an output file, which will be filled with an HTML page suitable to view your whole program's output as well as the input files.

Example

### CODE REVIEW

**Comment by pcworld** [comment]
This looks great. Only tested it functionality-wise, did not do a code review.

Some things I noticed:
- Either something is wrong with the encodings or these symbols do not exist in the specified font (FF and Chromium):  
  selection_212
- User output/input commands/input files should be escaped. (Yes, I actually have some output in final1 that is interpreted as an HTML tag: `Error, unknown command. Valid commands: nodes, edges, export, recommend <query>, quit.`)
- Maybe always underline input files (not only on hover) to make it clearer they can be clicked on.
- An "expand all" button might be useful.
- It's a bit confusing that a ".html" extension is added even if you specify one in the argument.

(I could fix some of them too if you don't want to.)
Edit 1: Added html extension issue

**Comment by pcworld** [comment]
If I let JUnit run only a single test, I get this exception:

```
java.lang.NullPointerException
    at test.OutputFileWriter.simpleName(OutputFileWriter.java:447)
    at test.OutputFileWriter.documentRun(OutputFileWriter.java:48)
    at test.InteractiveConsoleTest.sessionTest(InteractiveConsoleTest.java:601)
    at test.InteractiveConsoleTest.sessionTest(InteractiveConsoleTest.java:638)
    at final2.subtests.MoveTest.simpleMoveTest(MoveTest.java:63)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.FailOnTimeout$StatementThread.run(FailOnTimeout.java:74)
```

**Comment by jGleitz** [comment]
`NullPointerException` fixed. Forgot 2 lines.

**Comment by jGleitz** [comment]
> Either something is wrong with the encodings or these symbols do not exist in the specified font (FF and Chromium)

Neither. Encoding is set to `UTF-8` and the font, Ubuntu, sports the characters. But I think Google Fonts does not serve the whole font to make it smaller. I'll look into that.

**Comment by jGleitz** [comment]
> User output/input commands/input files should be escaped.

Absolutely. Added it in 0e33cbb, but in a quite conservative way. Would you give it a try?

**Comment by jGleitz** [comment]
> This looks great. […] did not do a code review.

`OutputFileWriter` is a bit messy. Also, the framework could use some refactoring since we introduced `Run`. But I don't have the time for that in the moment. And most likely, nobody will care for this repository after monday anymore (Although it actually might be quite useful for others, especially the next ones having to get the Programmieren-Übungsschein)

File: src/final2/subtests/InvalidCommandLineArgumentsTest.java [hunk 1]
```diff
@@ -12,7 +12,6 @@
  * @author Joshua Gleitze
  * @author Christian Hilden
  * @author Martin Löper
- * @author Roman Langrehr
```
Comment (RomanLangrehr) [inline src/final2/subtests/InvalidCommandLineArgumentsTest.java:15]: Why did you remove my author tag here?

Comment (jGleitz) [inline src/final2/subtests/InvalidCommandLineArgumentsTest.java:15]: Definitely not on purpose! Must have happened while merging, I'm sorry! Put it back in c55208c

**Comment by jGleitz** [comment]
We now get the Ubuntu Font from my server, files.joshuagleitze.de.

Advantages:
- The HTML output supports almost full `UTF-8`.

Disadvantages:
- People not having the Ubuntu font family installed will likely recognize longer page loading, as larger files have to be downloaded from my server.

Note that people that don't have the Ubuntu font family and are not connected to the internet will still have character issues.

Somebody please test this, I have Ubuntu installed so I can't do it.

**Comment by RomanLangrehr** [comment]
I'll test this soon, when my eclipse is responding again...

**Comment by jGleitz** [comment]
> It's a bit confusing that a ".html" extension is added even if you specify one in the argument

I thought that some inexperienced Windows users might not know how to open the file if it does not have the `.html` extension. But I see that it's irritating. You think it would be better not to add the extension?

**Comment by RomanLangrehr** [comment]
Maybe I used this feature wrong, but with the arguments:
image
I get the following message:
image

**Comment by jGleitz** [comment]
@RomanLangrehr You are not allowed to use any spaces in the arguments! (There is one space before and after the `=` of `outputFile`)

**Comment by RomanLangrehr** [comment]
Oh, soory...

**Comment by jGleitz** [comment]
@RomanLangrehr Quote your path if it contains spaces: `-DoutputFile="/home/josh/Mit Leerzeichen/%testname"`works

**Comment by RomanLangrehr** [comment]
I did this, but the generated page has some problems with chraacters...
image

**Comment by jGleitz** [comment]
In Eclipse, if you hit Edit > Set Encoding…, which encoding is used?

**Comment by jGleitz** [comment]
@RomanLangrehr try again with 9391dfe, please

**Comment by RomanLangrehr** [comment]
image

By the way: The error messages look pretty nice. Luckily my program isn't working at the moment ;)

**Comment by jGleitz** [comment]
> The error messages look pretty nice.

Yeah, I think so too. We got pretty far with this framework.

> Cp1252

We have encoding issues here. I use `UTF-8` and would advise you to do the same (You can set this solely for your JUnit-KIT project). Nevertheless, please try if the HTML looks good for you using your momentary encoding with 9391dfe

**Comment by pcworld** [comment]
> I thought that some inexperienced Windows users might not know how to open the file if it does not have the .html extension. But I see that it's irritating. You think it would be better not to add the extension?

Maybe introduce some simple `if (!str.endsWith(".html") || !str.endsWith(".htm"))` check

**Comment by RomanLangrehr** [comment]
@jGleitz Looks exactly as the page before. Sorry...

**Comment by jGleitz** [comment]
@RomanLangrehr: Yeah, I wasn't really escaping, my bad. Would you try again?

**Comment by jGleitz** [comment]
> Maybe introduce some simple if (!str.endsWith(".html") || !str.endsWith(".htm")) check

We might, but wouldn't this cause even more confusion? The momentary behaviour is consitent, at least.

**Comment by RomanLangrehr** [comment]
The page looks still the same... Sorry.

**Comment by pcworld** [comment]
I'm sure there are UTF-8 characters that look the same and are more common, i.e. exist in much more fonts. This would also avoid the need for loading a specific font.

> We might, but wouldn't this cause even more confusion? The momentary behaviour is consitent, at least.

I'd find it less confusing than always adding a file extension to a parameter that is already called "outputFile".

**Comment by jGleitz** [comment]
> I'd find it less confusing than always adding a file extension to a parameter that is already called "outputFile".

Alright. Added.

**Comment by jGleitz** [comment]
> I'm sure there are UTF-8 characters that look the same and are more common

There are, but they don't look as nice ;)

> This would also avoid the need for loading a specific font.

I'd load it anyway. I don't want to think of font configurations that look good and sport all characters and font styles we need for at least 4 operation systems.

**Comment by pcworld** [comment]
I still don't see your symbols with your latest commit.
Edit: I'm on Ubuntu so it doesn't load your files. Maybe there exist different Ubuntu fonts with the same name.

**Comment by jGleitz** [comment]
@RomanLangrehr @pcworld would you try again with 31556e5? It should now be completly independent from the encoding you use on your machine.

**Comment by pcworld** [comment]
No luck, see the edit on my previous comment:

> I'm on Ubuntu so it doesn't load your files. Maybe there exist different Ubuntu fonts with the same name

**Comment by jGleitz** [comment]
> I'm on Ubuntu so it doesn't load your files. Maybe there exist different Ubuntu fonts with the same name

There don't. If you're on Ubuntu using a halfway decent version, you've got the version I have on my server. I am on Ubuntu and Firefox doesn't load the fonts from my server, too. Nevertheless, none of the commits did change the output I get rendered.

@pcworld Could you take a screenshot of how it looks for you after 31556e5?

**Comment by pcworld** [comment]
> Could you take a screenshot of how it looks for you

On 41a30e98d8bc05f938659b56b7867f919396180d:
selection_211

> NullPointerException fixed. Forgot 2 lines.

When running a single test (by first running all final2 tests and than right clicking on a single test and choose "Run"), it doesn't write anything to the specified output file.

**Comment by jGleitz** [comment]
> When running a single test (by first running all final2 tests and than right clicking on a single test and choose "Run"), it doesn't write anything to the specified output file.

The system I adopted requires a test run with `KITSuite`. This allows us not to clutter the code. But therefore it will only work on suites, in the moment.

**Comment by RomanLangrehr** [comment]
On my PC the page has still the corrupted chracters...

**Comment by jGleitz** [comment]
@pcworld: You're sure he uses a font from your machine? I'm starting to think that, for whatever reason, the font on my server does not contains these characters.

**Comment by pcworld** [comment]
> You're sure he uses a font from your machine?

Yes, when loading the file in Chromium's incognito mode (i.e. without caches), it shows no remote network requests in the dev console.

**Comment by jGleitz** [comment]
I'm quite in the dark here. 

I commited a version forcing the browser to download the font (At least I hope so). Would you give it a try? (In the moment, it's not only about getting these characters to work, but rather understanding, why in the world this is not working.

**Comment by pcworld** [comment]
It now downloads your fonts but still looks the same.

**Comment by jGleitz** [comment]
With be9d516, do you find any more issues?

**Comment by pcworld** [comment]
Works now.
As a little accessibility feature, it would be nice if there was a text success/failure indicator, or differently looking symbols. But that's rather nice to have and not really necessary, we can merge now IMO.

**Comment by jGleitz** [comment]
I added the result to the indicator's title, which should be sufficient.

**Comment by RomanLangrehr** [comment]
image

Looks finally fine.
I think we can merge now...

**Comment by RomanLangrehr** [comment]
image

Finally it looks fine....

**Comment by pcworld** [comment]
In case anyone is interested, here is a patch for a little bit more accessible version:

``` diff
diff --git a/src/test/OutputFileWriter.java b/src/test/OutputFileWriter.java
index 2eac871..67de210 100644
--- a/src/test/OutputFileWriter.java
+++ b/src/test/OutputFileWriter.java
@@ -118,7 +118,8 @@ public class OutputFileWriter {
                    for (TestMethod testMethod : classMap.values()) {
                        String indicator = "<span class=\"successIndicator "
                                + ((testMethod.success) ? "succeeded" : "failed") + "\" title=\""
-                               + ((testMethod.success) ? "success" : "fail") + "\">" + CIRCLE_SVG + "</span>";
+                               + ((testMethod.success) ? "success" : "fail") + "\">"
+                               + ((testMethod.success) ? "✔" : "<span style=\"color: red;\">✘</span>") + "</span>";
                        write(new String[] {
                                "<div class=\"testMethod closed\" id=\"" + id + "\">",
                                "<h4 onclick=\"toggleMethod(" + id + ");\"><span class=\"opener\">" + TRIANGLE_SVG
```

selection_212
Anything against making this the default? To avoid encoding issues, we could just use HTML escapes, and I guess these two symbols should be found in most fonts.
Edit: Putting the success/failure indicator _before_ the method name would probably result in a better overview.
